### PR TITLE
Fix #125: Log warning when the amount of memory the JVM has is low.

### LINKED ISSUE #119: Running initial test run doesn't copy JVM system settings
### Summary

Running initial test run doesn't copy JVM system settings.

My tests require a JVM system setting (`-Dconfig.file=...`) to point to a configuration file for the environment where the tests are ran. Running `sbt -Dconfig.file=... test` runs fine. Running `sbt -Dconfig.file=... "project app" stryker` (since I'm using a multi-module structure) fails with:

    [info] *** 4 SUITES ABORTED ***
    [info] *** 14 TESTS FAILED ***
    [error] Failed tests:
    [error] 	hyperion.MeterAgentSpec
    [error] 	hyperion.DailyHistoryActorSpec
    [error] 	hyperion.RecentHistoryActorSpec
    [error] Error during tests:
    [error] 	hyperion.database.DatabaseActorSpec
    [error] 	hyperion.AppSettingsSpec
    [error] 	hyperion.rest.AppInfoServiceSpec
    [error] (Test / test) sbt.TestsFailedException: Tests unsuccessful
    [error] stryker4s.extensions.exceptions.InitialTestRunFailedException: Initial test run failed. Please make sure your tests pass before running Stryker4s.
    [error] stryker4s.extensions.exceptions.InitialTestRunFailedException: Initial test run failed. Please make sure your tests pass before running Stryker4s.
    [error] Use 'last' for the full log.

See [the build log](https://circleci.com/gh/mthmulders/hyperion/348) and [the source code repository](https://github.com/mthmulders/hyperion/tree/mutation-testing) for more details.

### Stryker config

```hocon
stryker4s {
  base-dir = "app"
  log-level = debug
  mutate = [
    "app/src/main/scala/**/*.scala"
  ]
  reporters = [
    console
  ]
}
```

### Stryker4s environment

```
stryker4s 0.1.0
```

### Your Environment
| software             | version(s) |
| -------------------- | ---------- |
| Scala version        | 2.12.8 |
| Build tool & version | SBT 1.2.8 | 
| Operating System     | MacOS Mojave (10.14.2) |

### LINKED ISSUE #125: Log a warning when Stryker4s is not given enough memory
We should log a warning when the JVM Stryker4s is running in is not given any extra memory, which we advise.

[Some info about how we could do that here](https://stackoverflow.com/questions/12807797/java-get-available-memory)

### PR DESCRIPTION
**Pull request by Wmaarts**
## Fixes #125

This logs a warning when Stryker4s is not given enough memory. Currently when max memory is below 2 GB.

### CODE REVIEW

**Comment by Wmaarts** [comment]
Right now the error message states to edit `SBT_OPTS` if you're using SBT. Once we implemented Maven we might want to edit this message to something that includes info for Maven users.

File: core/src/main/scala/stryker4s/Stryker4s.scala [hunk 1]
```diff
@@ -8,13 +9,24 @@ import stryker4s.run.report.Reporter
...
     val mutatedFiles = mutator.mutate(filesToMutate)
     val runResults = runner(mutatedFiles, fileCollector)
     reporter.report(runResults)
     ThresholdChecker.determineScoreStatus(runResults.mutationScore)
   }
+
+  private def preRunLogs(): Unit = {
+    // If JVM max memory is less than 2GB, log warning
+    if (!jvmHasEnoughMemory) {
+      warn("The JVM doesn't have a lot of memory assigned to it. It's wise to increase the maximum memory when running stryker.")
```
Comment (hugo-vrijswijk) [inline core/src/main/scala/stryker4s/Stryker4s.scala]: `doesn't have enough memory assigned to it` makes it sound like it's impossible to run Stryker4s (not stryker) with the current settings. Logging what we think is ideal (other than in the command) is probably best too. Oh, and sbt is lowercase.

File: core/src/test/scala/stryker4s/testutil/stubs/TestStryker4s.scala [hunk 2]
```diff
@@ -0,0 +1,15 @@
+package stryker4s.testutil.stubs
+
+import stryker4s.Stryker4s
+import stryker4s.config.Config
+import stryker4s.mutants.Mutator
+import stryker4s.mutants.findmutants.SourceCollector
+import stryker4s.run.MutantRunner
+import stryker4s.run.report.Reporter
+
+class TestStryker4s(jvmEnoughMemory: Boolean, fileCollector: SourceCollector, mutator: Mutator, runner: MutantRunner, reporter: Reporter)(
```
Comment (hugo-vrijswijk) [inline core/src/test/scala/stryker4s/testutil/stubs/TestStryker4s.scala]: This whole class can be removed (I think) by just using an anonymous class: 

```scala
val sut = new TestStryker4s(
  testSourceCollector,
  new Mutator(new MutantFinder(new MutantMatcher), new StatementTransformer, new MatchBuilder),
  testMutantRunner,
  testReporter) {
  
  override val jvmHasEnoughMemory: Boolean = false
}
```

File: core/src/test/scala/stryker4s/Stryker4sTest.scala [hunk 3]
```diff
@@ -48,5 +48,28 @@ class Stryker4sTest extends Stryker4sSuite {
...
+        jvmEnoughMemory = false,
+        testSourceCollector,
+        new Mutator(new MutantFinder(new MutantMatcher), new StatementTransformer, new MatchBuilder),
+        testMutantRunner,
+        testReporter
+      )
+
+      sut.run()
+
+      "The JVM doesn't have a lot of memory assigned to it. " +
```
Comment (hugo-vrijswijk) [inline core/src/test/scala/stryker4s/Stryker4sTest.scala]: We should also test if the JVM does have enough memory, that the warning is __not__ logged

File: core/src/main/scala/stryker4s/Stryker4s.scala [hunk 4]
```diff
@@ -8,13 +9,24 @@ import stryker4s.run.report.Reporter
...
+      warn("This can be done in SBT by adding setting an environment variable: SBT_OPTS=\"-XX:+CMSClassUnloadingEnabled -XX:MaxPermSize=4G -Xmx4G\" ")
+    }
+  }
+  protected def jvmHasEnoughMemory: Boolean =
+    (Runtime.getRuntime.maxMemory / 1024 / 1024) >= 1820
```
Comment (hugo-vrijswijk) [inline core/src/main/scala/stryker4s/Stryker4s.scala]: Magic numbers 🔮?

File: core/src/main/scala/stryker4s/Stryker4s.scala [hunk 5]
```diff
@@ -8,13 +9,24 @@ import stryker4s.run.report.Reporter
...
+    preRunLogs()
     val filesToMutate = fileCollector.collectFilesToMutate()
     val mutatedFiles = mutator.mutate(filesToMutate)
     val runResults = runner(mutatedFiles, fileCollector)
     reporter.report(runResults)
     ThresholdChecker.determineScoreStatus(runResults.mutationScore)
   }
+
+  private def preRunLogs(): Unit = {
+    // If JVM max memory is less than 2GB, log warning
```
Comment (hugo-vrijswijk) [inline core/src/main/scala/stryker4s/Stryker4s.scala]: Comment should probably say something about why we want more than 2 gigs of memory, not just what the code does

**Pull request merged**